File: src/cdp_capture.py
"""
CAS Cap - CDP経由でEdgeにアタッチし、ブラウザ操作＋スクリーンショットを取得する

前提:
  1. Edgeを全プロセス終了
  2. scripts/start_edge.ps1 でデバッグポート付きEdgeを起動
  3. exeからアプリを起動（Edgeにタブが追加される）
  4. このスクリプトを実行
"""

import asyncio
import argparse
import sys
from datetime import datetime
from pathlib import Path

import yaml
from playwright.async_api import async_playwright


def load_config(config_path: str = "config/config.yaml") -> dict:
    """設定ファイルを読み込む"""
    path = Path(config_path)
    if not path.exists():
        print(f"設定ファイルが見つかりません: {config_path}")
        print("config/config.sample.yaml をコピーして config/config.yaml を作成してください")
        sys.exit(1)
    with open(path, encoding="utf-8") as f:
        return yaml.safe_load(f)


async def list_tabs(cdp_url: str):
    """接続中のEdgeのタブ一覧を表示する"""
    async with async_playwright() as p:
        browser = await p.chromium.connect_over_cdp(cdp_url)
        print(f"接続成功。コンテキスト数: {len(browser.contexts)}")
        for ctx in browser.contexts:
            for i, page in enumerate(ctx.pages):
                print(f"  [{i}] {page.title}")
                print(f"      {page.url}")
        browser.close()


async def find_target_page(browser, target_url_keyword: str):
    """URLキーワードに一致するタブを探す"""
    for ctx in browser.contexts:
        for page in ctx.pages:
            if target_url_keyword in page.url:
                return page
    return None


async def capture(config: dict):
    """メインのキャプチャ処理"""
    cdp_url = config.get("cdp_url", "http://localhost:9222")
    target_keyword = config.get("target_url_keyword", "")
    output_dir = Path(config.get("output_dir", "output"))
    full_page = config.get("full_page", True)
    wait_state = config.get("wait_state", "networkidle")
    pages_config = config.get("pages", [])

    output_dir.mkdir(parents=True, exist_ok=True)
    timestamp = datetime.now().strftime("%Y%m%d_%H%M%S")

    async with async_playwright() as p:
        browser = await p.chromium.connect_over_cdp(cdp_url)
        print(f"Edge接続成功")

        # 対象タブを探す
        target_page = await find_target_page(browser, target_keyword)
        if target_page is None:
            print(f"対象タブが見つかりません (keyword: {target_keyword})")
            print("タブ一覧:")
            for ctx in browser.contexts:
                for page in ctx.pages:
                    print(f"  {page.url}")
            browser.close()
            sys.exit(1)

        print(f"対象タブ: {target_page.title}")
        await target_page.wait_for_load_state(wait_state)

        if not pages_config:
            # ページ設定がなければ現在の画面をキャプチャ
            filename = output_dir / f"capture_{timestamp}.png"
            await target_page.screenshot(path=str(filename), full_page=full_page)
            print(f"キャプチャ保存: {filename}")
        else:
            # 複数画面を巡回キャプチャ
            for i, page_conf in enumerate(pages_config):
                name = page_conf.get("name", f"page_{i:03d}")
                actions = page_conf.get("actions", [])

                # アクション実行
                for action in actions:
                    await execute_action(target_page, action)

                await target_page.wait_for_load_state(wait_state)

                filename = output_dir / f"{timestamp}_{name}.png"
                await target_page.screenshot(path=str(filename), full_page=full_page)
                print(f"キャプチャ保存: {filename}")

        browser.close()
        print("完了")


async def execute_action(page, action: dict):
    """ページ上でアクションを実行する"""
    action_type = action.get("type")

    if action_type == "click":
        selector = action["selector"]
        await page.click(selector)

    elif action_type == "fill":
        selector = action["selector"]
        value = action["value"]
        await page.fill(selector, value)

    elif action_type == "wait":
        selector = action.get("selector")
        timeout = action.get("timeout", 5000)
        if selector:
            await page.wait_for_selector(selector, timeout=timeout)
        else:
            await page.wait_for_timeout(timeout)

    elif action_type == "goto":
        url = action["url"]
        await page.goto(url)

    elif action_type == "select":
        selector = action["selector"]
        value = action["value"]
        await page.select_option(selector, value)

    elif action_type == "keyboard":
        key = action["key"]
        await page.keyboard.press(key)

    else:
        print(f"  未知のアクション: {action_type}")


def main():
    parser = argparse.ArgumentParser(description="CAS Cap - CDP経由ブラウザキャプチャツール")
    parser.add_argument("-c", "--config", default="config/config.yaml", help="設定ファイルパス")
    parser.add_argument("--list", action="store_true", help="タブ一覧を表示")
    parser.add_argument("--cdp-url", default=None, help="CDPのURL (デフォルト: config参照)")
    args = parser.parse_args()

    if args.list:
        cdp_url = args.cdp_url or "http://localhost:9222"
        asyncio.run(list_tabs(cdp_url))
    else:
        config = load_config(args.config)
        if args.cdp_url:
            config["cdp_url"] = args.cdp_url
        asyncio.run(capture(config))


if __name__ == "__main__":
    main()

File: js/cdp_capture.js
#!/usr/bin/env node
/*
 * CAS Cap (JavaScript版) - CDP経由でEdgeにアタッチし、ブラウザ操作＋スクリーンショットを取得する
 *
 * 前提:
 *   1. Edgeを全プロセス終了
 *   2. scripts/start_edge.ps1 でデバッグポート付きEdgeを起動
 *   3. exeからアプリを起動（Edgeにタブが追加される）
 *   4. このスクリプトを実行: node js/cdp_capture.js
 */

"use strict";

const fs = require("fs");
const path = require("path");
const { chromium } = require("playwright");

/** 設定ファイル（JSON）を読み込む */
function loadConfig(configPath) {
  if (!fs.existsSync(configPath)) {
    console.error(`設定ファイルが見つかりません: ${configPath}`);
    console.error("config/config.sample.json をコピーして config/config.json を作成してください");
    process.exit(1);
  }
  return JSON.parse(fs.readFileSync(configPath, "utf-8"));
}

/** YYYYMMDD_HHMMSS 形式のタイムスタンプ */
function timestamp() {
  const d = new Date();
  const p = (n) => String(n).padStart(2, "0");
  return (
    `${d.getFullYear()}${p(d.getMonth() + 1)}${p(d.getDate())}_` +
    `${p(d.getHours())}${p(d.getMinutes())}${p(d.getSeconds())}`
  );
}

/** 接続中のEdgeのタブ一覧を表示する */
async function listTabs(cdpUrl) {
  const browser = await chromium.connectOverCDP(cdpUrl);
  const contexts = browser.contexts();
  console.log(`接続成功。コンテキスト数: ${contexts.length}`);
  for (const ctx of contexts) {
    const pages = ctx.pages();
    for (let i = 0; i < pages.length; i++) {
      console.log(`  [${i}] ${await pages[i].title()}`);
      console.log(`      ${pages[i].url()}`);
    }
  }
  await browser.close();
}

/** URLキーワードに一致するタブを探す */
function findTargetPage(browser, targetKeyword) {
  for (const ctx of browser.contexts()) {
    for (const page of ctx.pages()) {
      if (page.url().includes(targetKeyword)) {
        return page;
      }
    }
  }
  return null;
}

/** ページ上でアクションを実行する */
async function executeAction(page, action) {
  switch (action.type) {
    case "click":
      await page.click(action.selector);
      break;
    case "fill":
      await page.fill(action.selector, action.value);
      break;
    case "wait":
      if (action.selector) {
        await page.waitForSelector(action.selector, { timeout: action.timeout ?? 5000 });
      } else {
        await page.waitForTimeout(action.timeout ?? 5000);
      }
      break;
    case "goto":
      await page.goto(action.url);
      break;
    case "select":
      await page.selectOption(action.selector, action.value);
      break;
    case "keyboard":
      await page.keyboard.press(action.key);
      break;
    default:
      console.log(`  未知のアクション: ${action.type}`);
  }
}

/** メインのキャプチャ処理 */
async function capture(config) {
  const cdpUrl = config.cdp_url ?? "http://localhost:9222";
  const targetKeyword = config.target_url_keyword ?? "";
  const outputDir = config.output_dir ?? "output";
  const fullPage = config.full_page ?? true;
  const waitState = config.wait_state ?? "networkidle";
  const pagesConfig = config.pages ?? [];

  fs.mkdirSync(outputDir, { recursive: true });
  const ts = timestamp();

  const browser = await chromium.connectOverCDP(cdpUrl);
  console.log("Edge接続成功");

  const targetPage = findTargetPage(browser, targetKeyword);
  if (targetPage === null) {
    console.error(`対象タブが見つかりません (keyword: ${targetKeyword})`);
    console.error("タブ一覧:");
    for (const ctx of browser.contexts()) {
      for (const page of ctx.pages()) {
        console.error(`  ${page.url()}`);
      }
    }
    await browser.close();
    process.exit(1);
  }

  console.log(`対象タブ: ${await targetPage.title()}`);
  await targetPage.waitForLoadState(waitState);

  if (pagesConfig.length === 0) {
    // ページ設定がなければ現在の画面をキャプチャ
    const filename = path.join(outputDir, `capture_${ts}.png`);
    await targetPage.screenshot({ path: filename, fullPage });
    console.log(`キャプチャ保存: ${filename}`);
  } else {
    // 複数画面を巡回キャプチャ
    for (let i = 0; i < pagesConfig.length; i++) {
      const pageConf = pagesConfig[i];
      const name = pageConf.name ?? `page_${String(i).padStart(3, "0")}`;
      const actions = pageConf.actions ?? [];

      for (const action of actions) {
        await executeAction(targetPage, action);
      }

      await targetPage.waitForLoadState(waitState);

      const filename = path.join(outputDir, `${ts}_${name}.png`);
      await targetPage.screenshot({ path: filename, fullPage });
      console.log(`キャプチャ保存: ${filename}`);
    }
  }

  await browser.close();
  console.log("完了");
}

/** コマンドライン引数をパースする */
function parseArgs(argv) {
  const args = { config: "config/config.json", list: false, cdpUrl: null };
  for (let i = 0; i < argv.length; i++) {
    const a = argv[i];
    if (a === "-c" || a === "--config") args.config = argv[++i];
    else if (a === "--list") args.list = true;
    else if (a === "--cdp-url") args.cdpUrl = argv[++i];
    else if (a === "-h" || a === "--help") {
      console.log("使い方: node cdp_capture.js [-c config.json] [--list] [--cdp-url URL]");
      process.exit(0);
    }
  }
  return args;
}

async function main() {
  const args = parseArgs(process.argv.slice(2));

  if (args.list) {
    const cdpUrl = args.cdpUrl ?? "http://localhost:9222";
    await listTabs(cdpUrl);
  } else {
    const config = loadConfig(args.config);
    if (args.cdpUrl) config.cdp_url = args.cdpUrl;
    await capture(config);
  }
}

main().catch((err) => {
  console.error(`エラー: ${err.message}`);
  process.exit(1);
});
